sympy/polys/matrices/tests/test_xxm.py

```diff
@@ -862,3 +862,126 @@ def test_XXM_lll(DM):
     assert M.lll() == M_lll
     assert M.lll_transform() == (M_lll, T)
     assert T.matmul(M) == M_lll
+
+
+@pytest.mark.parametrize('DM', DMQ_all)
+def test_XXM_qr_mixed_signs(DM):
+    lol = [[QQ(1), QQ(-2)], [QQ(-3), QQ(4)]]
+    A = DM(lol)
+    Q, R = A.qr()
+    assert Q.matmul(R) == A
+
+
+@pytest.mark.parametrize('DM', DMQ_all)
+def test_XXM_qr_large_matrix(DM):
+    lol = [[QQ(i + j) for j in range(10)] for i in range(10)]
+    A = DM(lol)
+    Q, R = A.qr()
+    assert Q.matmul(R) == A
+
+
+@pytest.mark.parametrize('DM', DMQ_all)
+def test_XXM_qr_identity_matrix(DM):
+    T = type(DM([[0]]))
+    A = T.eye(3, QQ)
+    Q, R = A.qr()
+    assert Q == A
+    assert R == A
+
+
+@pytest.mark.parametrize('DM', DMQ_all)
+def test_XXM_qr_square_matrix(DM):
+    lol = [[QQ(3), QQ(1)], [QQ(4), QQ(3)]]
+    A = DM(lol)
+    Q, R = A.qr()
+    assert Q.shape == (2, 2)
+    assert R.shape == (2, 2)
+    assert Q.matmul(R) == A
+
+
+@pytest.mark.parametrize('DM', DMQ_all)
+def test_XXM_qr_matrix_with_zero_columns(DM):
+    lol = [[QQ(3), QQ(0)], [QQ(4), QQ(0)]]
+    A = DM(lol)
+    Q, R = A.qr()
+    assert Q.shape == (2, 2)
+    assert R.shape == (2, 2)
+    assert Q.matmul(R) == A
```

Comment:
These tests should also check that `Q` is orthogonal and that `R` is upper triangular.

Reply:
Should orthogonality be manually asserted in the test or I should implement a `is_orthogonal` method in various classes (DDM, SDM) etc ?? since there is a `is_upper` method in those classes ?

Reply:
I think just something like `(Q*Q.transpose()).is_diagonal` is fine. Actually I see that this has `(Q.transpose()*Q).is_diagonal` instead. The docstring should make clear that this is a bit different from what `QRdecomposition` does because we don't want to make the orthogonal vectors into unit vectors.